### PR TITLE
Q&A RAG pipeline with MongoDB, Hugging Face and Llama3

### PR DESCRIPTION
**Pull request by RichmondAlake**
# What does this PR do?

**Description of Changes**: This update introduces an integrated pipeline using Retrieval-Augmented Generation (RAG) with MongoDB and Hugging Face's open-source Llama3 model for advanced question answering systems. The implementation covers the setup of a MongoDB database, data ingestion, vector search, and leveraging Llama3 for generating responses. These modifications aim to showcase how combined database and NLP technologies can improve data retrieval and processing capabilities for complex queries.

Motivation and Context: This change was motivated by the need to demonstrate practical applications of combining cutting-edge NLP models with modern database systems for educational and development purposes in the fields of AI and data science.

Dependencies:

- datasets for loading and processing the data.
- pandas for data manipulation.
- pymongo for interacting with MongoDB.
- sentence_transformers for generating text embeddings.
- transformers for accessing pre-trained models from Hugging Face.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X ] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb [hunk 1]
```diff
@@ -0,0 +1,1351 @@
...
+        "- Steps include dataset preparation, database setup, data ingestion, and query processing.\n",
+        "- Detailed guidance on setting up MongoDB collections and vector search indexes.\n",
+        "- Integration with the Llama3 model from Hugging Face for answering complex queries.\n",
+        "\n",
+        "Follow the following instruction to set up a MongoDB database and enable vector search:\n",
+        "1. [Register a free Atlas account](https://account.mongodb.com/account/register?utm_campaign=devrel&utm_source=community&utm_medium=cta&utm_content=GitHub%20Cookbook&utm_term=richmond.alake)\n",
+        " or sign in to your existing Atlas account.\n",
+        "\n",
+        "2. [Follow the instructions](https://www.mongodb.com/docs/atlas/tutorial/deploy-free-tier-cluster/)\n",
+        " (select Atlas UI as the procedure)  to deploy your first cluster.\n",
```
Comment (HamidShojanazeri) [inline recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb:28]: @RichmondAlake can we pls have a brief explanation what does cluster mean here.

File: recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb [hunk 2]
```diff
@@ -0,0 +1,1351 @@
...
+      "metadata": {
+        "id": "uRvntN3jO8j1"
+      },
+      "outputs": [],
+      "source": [
+        "# Load Dataset\n",
+        "from datasets import load_dataset\n",
+        "import pandas as pd\n",
+        "\n",
+        "# Make sure you have an Hugging Face token(HF_TOKEN) in your development environemnt before runing the code below\n",
```
Comment (HamidShojanazeri) [inline recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb:85]: can we pls have a place holder for HF token to make sure users know how to set/ enforce it not to forget

File: recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb [hunk 3]
```diff
@@ -0,0 +1,1351 @@
...
+        "- Integration with the Llama3 model from Hugging Face for answering complex queries.\n",
+        "\n",
+        "Follow the following instruction to set up a MongoDB database and enable vector search:\n",
+        "1. [Register a free Atlas account](https://account.mongodb.com/account/register?utm_campaign=devrel&utm_source=community&utm_medium=cta&utm_content=GitHub%20Cookbook&utm_term=richmond.alake)\n",
+        " or sign in to your existing Atlas account.\n",
+        "\n",
+        "2. [Follow the instructions](https://www.mongodb.com/docs/atlas/tutorial/deploy-free-tier-cluster/)\n",
+        " (select Atlas UI as the procedure)  to deploy your first cluster.\n",
+        "\n",
+        "3. Create the database: `movies`.\n",
```
Comment (HamidShojanazeri) [inline recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb:30]: is this name correct here, it seems we are loading a dataset of papers.

**Review by HamidShojanazeri — COMMENTED** [review]
Thanks @RichmondAlake for the PR! just left few minor comments.

**Comment by RichmondAlake** [comment]
@HamidShojanazeri Thanks for reviewing very quickly.

I have made the modifications and suggested changes.

Richmond

File: recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb [hunk 4]
```diff
@@ -0,0 +1,1354 @@
...
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "Connection to MongoDB successful\n"
+          ]
+        }
+      ],
+      "source": [
+        "import pymongo\n",
+        "from google.colab import userdata\n",
```
Comment (jeffxtang) [inline recipes/use_cases/chatbots/RAG_chatbot/vectorstore/mongodb/rag_mongodb_llama3_huggingface_open_source.ipynb:649]: Running this notebook on local jupyter notebook I got this error:
```
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[5], line 2
      1 import pymongo
----> 2 from google.colab import userdata
      4 def get_mongo_client(mongo_uri):
      5   """Establish connection to the MongoDB."""

ModuleNotFoundError: No module named 'google.colab'
```

**Comment by RichmondAlake** [comment]
@jeffxtang Thanks for pointing out these issues.

I have addressed them and added screenshots where appropriate.

**Comment by jeffxtang** [comment]
Thanks @RichmondAlake for the changes and the screenshots. I tried following steps under Atlas UI (there're 10!) but got lost - I already logged in and completed the survey then I clicked Try Free and it asked me to Sign up/in again - with Sign in I saw the Overview - Create a deployment page, but the instructions with 10 steps page was no longer visible... 
The RAG and Llama 3 parts look good, but I'm afraid many users will probably be stuck in the Mongodb setup... 

Is there another simpler way to to showing MongoDB for this use case? A while ago I downloaded some Mongodb locally on Mac to use it with Llama 2, which worked pretty easily (the setup process and integration).

I can approve this PR if you're fine with the current DB setup process.

**Comment by RichmondAlake** [comment]
@jeffxtang thanks for your comment.

Please move forward with merging the PR.

I will work on the experience with the docs with our internal team.

Thanks for the review

**Review by jeffxtang — APPROVED** [review]

**Pull request merged**